File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/microservice/MicroServiceExchangeImpl.java
package org.babyfish.jimmer.sql.microservice;

import org.babyfish.jimmer.meta.ImmutableProp;
import org.babyfish.jimmer.runtime.ImmutableSpi;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.tuple.Tuple2;
import org.babyfish.jimmer.sql.common.AbstractTest;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.babyfish.jimmer.sql.model.JimmerModule;
import org.babyfish.jimmer.sql.runtime.ConnectionManager;
import org.babyfish.jimmer.sql.runtime.MicroServiceExchange;
import org.babyfish.jimmer.sql.runtime.MicroServiceExporter;

import java.sql.Connection;
import java.util.Collection;
import java.util.List;
import java.util.function.Function;

public class MicroServiceExchangeImpl implements MicroServiceExchange {

    private static final ConnectionManager CONNECTION_MANAGER =
            new ConnectionManager() {
                @SuppressWarnings("unchecked")
                @Override
                public <R> R execute(Function<Connection, R> block) {
                    R[] ref = (R[])new Object[1];
                    AbstractTest.jdbc(con -> {
                        ref[0] = block.apply(con);
                    });
                    return ref[0];
                }
            };

    private final JSqlClient orderClient =
            JSqlClient
                    .newBuilder()
                    .setEntityManager(JimmerModule.ENTITY_MANAGER)
                    .setConnectionManager(CONNECTION_MANAGER)
                    .setMicroServiceName("order-service")
                    .setMicroServiceExchange(this)
                    .build();

    private final JSqlClient orderItemClient =
            JSqlClient
                    .newBuilder()
                    .setEntityManager(JimmerModule.ENTITY_MANAGER)
                    .setConnectionManager(CONNECTION_MANAGER)
                    .setMicroServiceName("order-item-service")
                    .setMicroServiceExchange(this)
                    .build();

    private final JSqlClient productClient =
            JSqlClient
                    .newBuilder()
                    .setEntityManager(JimmerModule.ENTITY_MANAGER)
                    .setConnectionManager(CONNECTION_MANAGER)
                    .setMicroServiceName("product-service")
                    .setMicroServiceExchange(this)
                    .build();

    @SuppressWarnings("unchecked")
    @Override
    public List<ImmutableSpi> findByIds(
            String microServiceName,
            Collection<?> ids,
            Fetcher<?> fetcher
    ) {
        return new MicroServiceExporter(sqlClient(microServiceName))
                .findByIds(ids, fetcher);
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<Tuple2<Object, ImmutableSpi>> findByAssociatedIds(
            String microServiceName,
            ImmutableProp prop,
            Collection<?> targetIds,
            Fetcher<?> fetcher
    ) {
        return new MicroServiceExporter(sqlClient(microServiceName))
                .findByAssociatedIds(prop, targetIds, fetcher);
    }

    private JSqlClient sqlClient(String microServiceName) {
        switch (microServiceName) {
            case "order-service":
                return orderClient;
            case "order-item-service":
                return orderItemClient;
            case "product-service":
                return productClient;
            default:
                throw new IllegalArgumentException(
                        "Illegal microservice name \"" +
                                microServiceName +
                                "\""
                );
        }
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/microservice/MicroServiceExchangeImpl.kt
package org.babyfish.jimmer.sql.kt.microservice

import org.babyfish.jimmer.meta.ImmutableProp
import org.babyfish.jimmer.runtime.ImmutableSpi
import org.babyfish.jimmer.sql.ast.tuple.Tuple2
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.common.AbstractTest
import org.babyfish.jimmer.sql.kt.impl.KSqlClientImplementor
import org.babyfish.jimmer.sql.kt.newKSqlClient
import org.babyfish.jimmer.sql.runtime.ConnectionManager
import org.babyfish.jimmer.sql.runtime.MicroServiceExchange
import org.babyfish.jimmer.sql.runtime.MicroServiceExporter
import java.sql.Connection
import java.util.function.Function

class MicroServiceExchangeImpl() : MicroServiceExchange {

    private val orderClient = newKSqlClient {
        setConnectionManager(CONNECTION_MANAGER)
        setMicroServiceName("order-service")
        setMicroServiceExchange(this@MicroServiceExchangeImpl)
    }

    private val orderItemClient = newKSqlClient {
        setConnectionManager(CONNECTION_MANAGER)
        setMicroServiceName("order-item-service")
        setMicroServiceExchange(this@MicroServiceExchangeImpl)
    }

    private val productClient = newKSqlClient {
        setConnectionManager(CONNECTION_MANAGER)
        setMicroServiceName("product-service")
        setMicroServiceExchange(this@MicroServiceExchangeImpl)
    }

    override fun findByIds(
        microServiceName: String,
        ids: Collection<*>,
        fetcher: Fetcher<*>
    ): List<ImmutableSpi> {
        return MicroServiceExporter((sqlClient(microServiceName) as KSqlClientImplementor).javaClient)
            .findByIds(ids, fetcher)
    }

    override fun findByAssociatedIds(
        microServiceName: String,
        prop: ImmutableProp,
        targetIds: Collection<*>,
        fetcher: Fetcher<*>
    ): List<Tuple2<Any, ImmutableSpi>> {
        return MicroServiceExporter((sqlClient(microServiceName) as KSqlClientImplementor).javaClient)
            .findByAssociatedIds(prop, targetIds, fetcher)
    }

    private fun sqlClient(microServiceName: String): KSqlClient =
        when (microServiceName) {
            "order-service" -> orderClient
            "order-item-service" -> orderItemClient
            "product-service" -> productClient
            else -> throw IllegalArgumentException(
                "Illegal microservice name \"" +
                    microServiceName +
                    "\""
            )
        }

    companion object {

        @Suppress("UNCHECKED_CAST")
        private val CONNECTION_MANAGER: ConnectionManager = object : ConnectionManager {
            override fun <R> execute(con: Connection?, block: Function<Connection, R>): R {
                val ref = arrayOfNulls<Any>(1) as Array<R>
                if (con == null) {
                    AbstractTest.jdbc { ref[0] = block.apply(it) }
                } else {
                    ref[0] = block.apply(con)
                }
                return ref[0]
            }
        }
    }
}